File: apps/android/core/compat/src/main/kotlin/app/gauja/core/compat/FeatureGate.kt
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
package app.gauja.core.compat

import app.gauja.core.model.servers.ServerStatus
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class FeatureMetadata(
    val min: String,
    val max: String?,
    val endpoint: String,
    val note: String,
)

class FeatureGate(private val metadata: Map<String, FeatureMetadata>) {
    fun isSupported(featureId: String, version: ServerVersion?): Boolean {
        val feature = metadata[featureId] ?: return false
        val minimum = ServerVersion.parse(feature.min) ?: return false
        val maximum = feature.max?.let { ServerVersion.parse(it) ?: return false }
        return version != null && version >= minimum && (maximum == null || version <= maximum)
    }

    fun outsideSupportedRange(status: ServerStatus): Boolean {
        val version = ServerVersion.parse(status.version)
        return metadata.keys.none { isSupported(it, version) }
    }

    fun metadata(featureId: String): FeatureMetadata? = metadata[featureId]

    companion object {
        fun bundled(): FeatureGate =
            requireNotNull(FeatureGate::class.java.getResourceAsStream("/compat.json")).use {
                decode(it.bufferedReader().readText())
            }

        fun decode(json: String): FeatureGate =
            FeatureGate(Json { ignoreUnknownKeys = true }.decodeFromString(json))
    }
}

File: apps/ios/Packages/Compat/Sources/Compat/FeatureGate.swift
// SPDX-FileCopyrightText: 2026 Gauja contributors
// SPDX-License-Identifier: AGPL-3.0-or-later
import Foundation

public struct FeatureMetadata: Sendable, Decodable {
    public let min: String
    public let max: String?
    public let endpoint: String
    public let note: String
}

public struct FeatureGate: Sendable {
    private let features: [String: FeatureMetadata]
    public init(data: Data) throws { features = try JSONDecoder().decode([String: FeatureMetadata].self, from: data) }

    public static func bundled() throws -> Self {
        guard let url = Bundle.module.url(forResource: "compat", withExtension: "json") else {
            throw CocoaError(.fileNoSuchFile)
        }
        return try Self(data: Data(contentsOf: url))
    }

    public func metadata(_ featureID: String) -> FeatureMetadata? { features[featureID] }

    public func isSupported(_ featureID: String, version: ServerVersion?) -> Bool {
        guard let feature = features[featureID], let minimum = ServerVersion(feature.min),
            let version, version >= minimum
        else { return false }
        guard let maximumText = feature.max else { return true }
        guard let maximum = ServerVersion(maximumText) else { return false }
        return version <= maximum
    }
}
